torchx/schedulers/azure_batch_scheduler.py

```diff
@@ -0,0 +1,338 @@
+
+"""
+
+from dataclasses import dataclass
+from typing import Dict, List, Optional, TypedDict
+import yaml
+from torchx.schedulers.api import (
+    AppDryRunInfo,
+    DescribeAppResponse,
+    ListAppResponse,
+    Scheduler
+)
+from torchx.schedulers.ids import make_unique
+
+from torchx.specs.api import (
+    AppDef,
+    AppState,
+    macros,
+    runopts,
+    RetryPolicy,
+    Role,
+)
+
+from azure.batch import BatchServiceClient
+
+import azure.batch.models as batchmodels
+from azure.batch import BatchServiceClient
+from azure.batch.batch_auth import SharedKeyCredentials
+import azure.batch.models as batchmodels
+
+
+
+@dataclass
+class AzureBatchJob:
+    job_id: str
+    pool_id: str
+    job: batchmodels.JobAddParameter
+    tasks: List[batchmodels.TaskAddParameter]
+
+    def __str__(self) -> str:
+        return yaml.dump({"job": self.job, "tasks": self.tasks})
+
+    def __repr__(self) -> str:
+        return str(self)
+
+
+class AzureBatchOpts(TypedDict, total=False):
+    batch_container_registry: Optional[str]
+    batch_container_username: Optional[str]
+    batch_container_password: Optional[str]
+    batch_pool_id: Optional[str]
+
+    
+JOB_STATE: Dict[str, AppState] = {
+    # TODO: instead of state to state mapping, use Job transition states (prev+current) to capture status
+    batchmodels.JobState.active: AppState.RUNNING,
+    batchmodels.JobState.completed: AppState.SUCCEEDED,
+    batchmodels.JobState.deleting: AppState.UNKNOWN,
+    batchmodels.JobState.disabled: AppState.UNKNOWN,
+    batchmodels.JobState.disabling: AppState.UNKNOWN,
+    batchmodels.JobState.enabling: AppState.PENDING,
+    batchmodels.JobState.terminating: AppState.UNKNOWN,
+}
+
+def _as_metadata_model(name:str, value:str) -> batchmodels.MetadataItem:
+    # pyre-ignore[16]
+    return batchmodels.MetadataItem(name=name, value=value)
+
+class AzureBatchScheduler(Scheduler[AzureBatchOpts]):
+    """
+    AzureBatchScheduler is a TorchX scheduling interface to Azure Batch.
+
+    .. code-block:: bash
+
+        $ pip install torchx
+        $ torchx run --scheduler azure_batch --scheduler_args image_repo=ghcr.io/pytorch/torchx --image alpine:latest --msg hello
+        azure_batch://torchx_user/1234
+        $ torchx describe azure_batch://torchx_user/1234
+        ...
+
+    **Config Options**
+
+    .. runopts::
+        class: torchx.schedulers.azure_batch_scheduler.create_scheduler
+
+
+    **Compatibility**
+
+    .. compatibility::
+        type: scheduler
+        features:
+            cancel: true
+            logs: false
+            distributed: true
+            describe: true
+            workspaces: false
+            mounts: false
+            elasticity: false
+    """
+
+    def __init__(
+        self,
+        session_name: str,
+        batch_account_name: Optional[str],
+        batch_account_key: Optional[str],
+        batch_account_url: Optional[str],
```

Comment:
why do these have to be taken in the constructor arguments (versus via the runoptions)? Looks like they are used when building the client in `_build_client()` and the client is create on-demand in the `schedule` API (rather than it being created once and being cached somewhere).

It looks like the real underlying reason why you added these to the constructor arguments is b/c the runcfg is not passed to the other APIs (e.g. list, describe, etc). Why not make the runcfg available to the other APIs instead of splitting scheduler configs between constructor args and runoptions?